lib/trace/trace_parser_loop.cpp

```diff
@@ -87,34 +79,23 @@ LastFrameLoopParser::parse_call(void)
 
     /* Restart last frame when looping is requested. */
     if (call) {
-        if (call->flags & trace::CALL_FLAG_END_FRAME) {
-            lastFrameBegin = lastFrameEnd;
-            parser->getBookmark(lastFrameEnd);
+        if (starts_new_frame) {
+            for (auto c : lastFrameCalls)
+                delete c;
+            lastFrameCalls.clear();
         }
+
+        call->reuse_call = true;
+        lastFrameCalls.push_back(call);
+        starts_new_frame = call->flags & trace::CALL_FLAG_END_FRAME;
+
+        lastFrameIterator = lastFrameCalls.begin();
+
         return call;
     }
 
     /* else */
     if (loopCount) {
-        if (lastFrameCalls.empty()) {
-            /* initialize vector with calls for the last frame, so we
-             * don't have to re-parse when looping. */
-            parser->setBookmark(lastFrameBegin);
-            while (true) {
-                call = parser->parse_call();
-                /* prevent call from being cleaned up after retrace.
-                 * It sill be retraced manny times. */
-                call->reuse_call = true;
-                lastFrameCalls.push_back(call);
-                ParseBookmark current;
-                parser->getBookmark(current);
-                if (current.next_call_no == lastFrameEnd.next_call_no) {
-                    /* end of frame */
-                    lastFrameIterator = lastFrameCalls.begin();
-                    break;
-                }
-            }
-        }
         call = *lastFrameIterator;
         ++lastFrameIterator;
         if (lastFrameIterator == lastFrameCalls.end()) {
```

Comment:
I guess this  deserves a `Closes: #729` in the commit message, right?

Reply:
Indeed, thanks!